### DISCUSSION TITLE
Is there any way to use the "disableDefinition" equivalent on the implementation side of the extension?

### DISCUSSION THREAD

**Discussion by yaegassy** [neoclide/coc.nvim#3030] [Q&A] [answered]
The `LanguageClientOptions` interface in coc.nvim allows for `disableCompletion`, `disableSnippetCompletion`, `disableDiagnostics`, and more.

For example, if you use two language servers (extension), you may want to disable the "definition" of only one of them. (Both provide a definition).

**Comment by fannheyward** [comment, marked as answer]
No `disableDefinition` API yet, you can use `provideDefinition` middleware to return null/undefined result if you want to disable it.

**Reply by yaegassy** [reply]
Thanks @fannheyward.  I just tried it, but void doesn't seem to work. :(

```typescript
  const clientOptions: LanguageClientOptions = {
    // ...snip
    middleware: {
      // ...snip
      provideDefinition: (
        document: TextDocument,
        position: Position,
        token: CancellationToken,
        next: ProvideDefinitionSignature
      ) => {
        next(document, position, token);
      },
    },
  };
```

**tsserver error**

```
src/index.ts|332 col 7 error| [tsserver 2322] Type '(document: TextDocument, position: Position, token: CancellationToken, next: ProvideDefinitionSignature) => void' is not assignable to type '(this: void, document: TextDocument, position: Position, token: CancellationToken, next: ProvideDefinitionSignature) => ProviderResult<...>'. [E]
```

**Reply by fannheyward** [reply]
@yaegassy just `return`, no need to call `next` anymore because you didn't want it.

**Reply by fannheyward** [reply]
https://github.com/fannheyward/coc-deno/blob/master/client/src/extension.ts#L101

**Reply by yaegassy** [reply]
I am thinking of adding configuration to enable/disable it.

I'll refer to the coc-deno code! Thank you very much.

**Reply by fannheyward** [reply]
reruns null/undefined, not void, my mistake. https://github.com/neoclide/coc.nvim/blob/master/typings/index.d.ts#L2999-L3014

**Reply by yaegassy** [reply]
I was able to behave as I expected. Thanks @fannheyward.

**(For future searchers)**

```typescript
  const clientOptions: LanguageClientOptions = {
    // ...snip
    middleware: {
      // ...snip
      provideDefinition: async (
        document: TextDocument,
        position: Position,
        token: CancellationToken,
        next: ProvideDefinitionSignature
      ) => {
        // ---- configuration ----
        if (disableDefinition) {
          return;
        }
        const def = await next(document, position, token);
        return def;
      },
    },
  };
```
